src/transformStateWithClones.js

```diff
@@ -8,6 +8,40 @@
  */
 function transformStateWithClones(state, actions) {
   // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    let nextState;
+
+    switch (action.type) {
+      case 'clear':
+        nextState = {};
+        break;
+
+      case 'addProperties':
+        nextState = {
+          ...currentState, ...action.extraData,
+        };
+        break;
+
+      case 'removeProperties':
+        nextState = { ...currentState };
+
+        for (const key of action.keysToRemove) {
+          delete nextState[key];
+        }
+        break;
+
+      default:
+        nextState = currentState;
+    }
+
+    stateHistory.push({ ...nextState });
+    currentState = { ...nextState };
```

Comment:
no need to use `nextState`. You can modify `currentState` and push its copy to the array